decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,210 @@
+package openlr.decoder.backtracking;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class BackTrackingGraph<T> {
+    private final List<Slice> slices;
+    private final Map<String, GraphEdge> edges;
+    private int currentSourceSliceIndex = -1;
+    private GraphEdge<T> graphEdge = null;
+    private GraphEdge<T> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    private int findNextNodeIndexIfCurrentIsRed(Slice sourceSlice) {
```

Comment:
Can you comment these methods to make clear what they do?

Reply:
comments are added

Reply:
Thanks!